### ISSUE TITLE
Define upgrade path for ComponentConfig

### ISSUE THREAD

**Issue by fabriziopandini** [kubernetes/kubeadm#1681]
kubeadm allow usage of component config for defining settings for kubelet and kube-proxy; component configs are stored in ConfigMaps and reused when joining nodes or doing upgrades.

However, as of today kubeadm does not yet define a procedure for upgrading component configs in the case such API change format e.g. change from `kube-proxy/config/v1alpha1` to `kube-proxy/config/v1alpha2`.

As an initial step, we should evaluate if this is the scope of kubeadm because it requires a certain degree of coordination across nodes (there should not be old components referring to a new/not supported component config), and this can't be achieved by kubeadm alone.
If this is the case, it should be documented what is the expected procedure higher-level tools/user's should take care of.

It is also important to keep in mind nuances of component config management for kubelet (1CM per version, copied locally) and kube-proxy (1CM, linked by all instances).

/cc @rosti

**Comment by rosti** [comment]
I was actually raising this exact question in the past few weeks with WG Component Standard, but so far no concrete decisions have been taken.
This same problem is faced, not only by kubeadm, but by other cluster provisioning tools too (like kops for instance). So I suppose that we need a broader WG Component Standard meeting with kubeadm and other cluster provisioning tools on that matter.
@mtaufen @stealthybox @sttts @justinsb WDYT?

**Comment by neolit123** [comment]
/assign @rosti 
xref https://github.com/kubernetes/enhancements/pull/1256/files

**Comment by xlgao-zju** [comment]
like what we discussed in https://github.com/kubernetes/kubeadm/issues/1582 
seems up-convert these component configs is out of kubeadm's scope...?

**Comment by neolit123** [comment]
yes, unless the component start exposing a conversion path for their managed ComponentConfig API / types.

**Comment by neolit123** [comment]
the kube-proxy v1alpha2 is planned for release in 1.31.
https://github.com/kubernetes/kubernetes/pull/121830
cc @SataQiu @pacoxu @carlory 

 kube-proxy has enabled the following as a migration plan, which is good:

https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/784-kube-proxy-component-config/README.md#upgrade--downgrade-strategy

```
/usr/local/bin/kube-proxy --config old-v1alpha1.yaml --write-config-to new-v1alpha2.yaml
```

since kubeadm runs kube-proxy as a container / DS we might have to jump a couple of additional hoops, but up-converting the already-in-cluster kube-proxy config to v1alpha2 seems doable.

- during upgrade apply if there is a kube-system/kube-proxy config CM:
  - download the config and save it somewhere locally
  - start a Job with the kube-proxy container, mounting the config location
  - call `/usr/local/bin/kube-proxy --config old-v1alpha1.yaml --write-config-to ...`
  - write the resulting config to the kube-system/kube-proxy config CM

i will experiment with this and provide a PR or if there are complications this can be a google doc first.
please add any comments if needed.

**Comment by neolit123** [comment]
looks like both kube-proxy and kube-scheduler support the --write-config-to flag. the kubelet does not yet
logged this ticket for it:
https://github.com/kubernetes/kubernetes/issues/125344

**Comment by neolit123** [comment]
in terms of the ticket here, IMO the upgrade path is:
- kubeadm calls the component binary to migrate the user config on disk and in cluster.
- kubeadm continues to accept the old CC config on `init` for N releases (to be discussed).
  - if the user passes an old config, kubeadm calls the component to convert it.
